File: 热门100/LRU.cpp
// 定义双向链表
struct DoubleLinkedList
{
  int key, val;
  DoubleLinkedList *prev;
  DoubleLinkedList *next;
  DoubleLinkedList() : key(0), val(0), prev(nullptr), next(nullptr) {}
  DoubleLinkedList(int _key, int _val) : key(_key), val(_val), prev(nullptr), next(nullptr) {}
};

class LRUCache
{
private:
  unordered_map<int, DoubleLinkedList *> cache;
  DoubleLinkedList *head;
  DoubleLinkedList *tail;
  int capacity;
  int size;
  void remove(DoubleLinkedList *node)
  {
    node->next->prev = node->prev;
    node->prev->next = node->next;
  }

  void addToHead(DoubleLinkedList *node)
  {
    node->next = head->next;
    node->prev = head;
    head->next->prev = node;
    head->next = node;
  }

  void moveToHead(DoubleLinkedList *node)
  {
    remove(node);
    addToHead(node);
  }

  DoubleLinkedList *removeTail()
  {
    DoubleLinkedList *node = tail->prev;
    remove(node);
    return node;
  }

public:
  LRUCache(int _capacity) : capacity(_capacity), size(0)
  {
    head = new DoubleLinkedList();
    tail = new DoubleLinkedList();
    head->next = tail;
    tail->prev = head;
  }

  int get(int key)
  {
    if (!cache.count(key)) // 不存在
    {
      return -1;
    }
    // 如果 key 存在，通过hash定位，再移到头部
    DoubleLinkedList *node = cache[key];
    moveToHead(node);
    return node->val;
  }

  void put(int key, int val)
  {
    // 如果没有，就新增
    if (!cache.count(key))
    {
      DoubleLinkedList *node = new DoubleLinkedList(key, val);
      addToHead(node);
      cache[key] = node;
      size++;

      if (size > capacity)
      {
        DoubleLinkedList *removed = removeTail();
        cache.erase(removed->key); // 从哈希表中移除
        delete removed;            // 防止内存泄漏
        size--;
      }
    }
    else
    {
      DoubleLinkedList *node = cache[key];
      moveToHead(node);
      node->val = val;
    }
  }
};

File: 剑指Offer/py实现/LRU.py
class DoubleLinkListNode:
    def __init__(self, key: int, val: int):
        self.key = key
        self.val = val
        self.prev = None
        self.next = None


class LRUCache:
    def __init__(self, capacity: int):
        self.cache = dict()
        self.head = DoubleLinkListNode(0, 0)
        self.tail = DoubleLinkListNode(0, 0)
        self.head.next = self.tail
        self.tail.next = self.head
        self.capacity = capacity
        self.size = 0

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1
        node = self.cache[key]
        self.__move_to_head(node)
        return node.val

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            node = self.cache[key]
            node.val = value
            self.__move_to_head(node)
        else:
            new_node = DoubleLinkListNode(key, value)
            self.__add_to_head(new_node)
            self.cache[key] = new_node
            self.size += 1

            if self.size > self.capacity:
                node = self.__remove_tail()
                self.cache.pop(node.key)
                self.size -= 1

    def __move_to_head(self, node):
        self.__remove_node(node)
        self.__add_to_head(node)

    def __remove_node(self, node: DoubleLinkListNode):
        node.prev.next = node.next
        node.next.prev = node.prev

    def __add_to_head(self, node: DoubleLinkListNode):
        node.next = self.head.next
        node.prev = self.head.prev
        self.head.next.prev = node
        self.head.next = node

    def __remove_tail(self):
        node = self.tail.prev
        self.__remove_node(node)
        return node
